### PR TITLE
Add newblobs

### PR DESCRIPTION
**Pull request by iwikal**
Please bikeshed the name. "newblob" is inscrutible unless you happen to know of
the FP concept of newtypes, and even then it's not obvious.

### CODE REVIEW

File: sylt-compiler/src/typechecker.rs [hunk 1]
```diff
@@ -1133,8 +1136,20 @@ impl TypeChecker {
...
+                    Type::Newblob(name, fields, _) => {
+                        errors.push(type_error!(
+                            self,
+                            span,
+                            TypeError::NewblobInstance { name: name.clone() }
+                        ));
+
+                        (name, fields)
+                    }
                     _ => unreachable!(),
```
Comment (FredTheDino) [inline sylt-compiler/src/typechecker.rs:1153]: This is a lie! I'll make an issue of it.

```
A :: enum
end

start :: fn do
    A {}
end
```
thread 'main' panicked at 'internal error: entered unreachable code', sylt-compiler/src/typechecker.rs:1153:26

File: sylt-compiler/src/typechecker.rs [hunk 2]
```diff
@@ -1383,7 +1397,7 @@ impl TypeChecker {
                     .collect(),
                 Box::new(self.inner_bake_type(ret, seen)),
             ),
-            Type::Blob(name, fields) => RuntimeType::Blob(
+            Type::Newblob(name, fields, _) | Type::Blob(name, fields) => RuntimeType::Blob(
```
Comment (FredTheDino) [inline sylt-compiler/src/typechecker.rs:1400]: This is used for type-errors, so it's a good idea to add a new field to `RuntimeType` and split this, it will make errors clearer. (`RuntimeType` is in sylt-common/src/ty.rs and there called `Type`)

**Review by FredTheDino — COMMENTED** [review]
Nice! I like the tests! It would be nice with a test with newblobs in different files with the same name, and seeing what happens when you unify those since it's a bit of an edge case in the error reporting.

Alternatives to newblob might be `unique`, `xblob`, `externalblob`, `externaltype`? Maybe one of those are good?

File: sylt-compiler/src/typechecker.rs [hunk 3]
```diff
@@ -1133,8 +1136,20 @@ impl TypeChecker {
...
+                let mut errors = Vec::new();
+
                 let (blob_name, blob_fields) = match self.find_type(blob_ty) {
                     Type::Blob(name, fields) => (name, fields),
+                    Type::Newblob(name, fields, _) => {
+                        errors.push(type_error!(
+                            self,
+                            span,
+                            TypeError::NewblobInstance { name: name.clone() }
+                        ));
```
[the remark this answers is not in the document]
Comment (iwikal) [reply sylt-compiler/src/typechecker.rs:1149]: I've been bothered by how, when there are lots of errors, the most relevant one is often at the top, forcing you to scroll the terminal to find it. Python stacktraces often seem to know to put the most relevant error last. I know stacktraces and compilation errors are very different problems to solve, but maybe there is some inspiration we can gather anyway? For now, I think you're right.

[the remark this answers is not in the document]
Comment (FredTheDino) [reply sylt-compiler/src/typechecker.rs:1149]: I find those errors a tad confusing, since I always read them the wrong way out of habit. But we can absolutely just reverse the list of errors before we print them out. It's a 1 line fix. If you want to do that, go for it!

File: sylt-compiler/src/typechecker.rs [hunk 4]
```diff
@@ -1756,6 +1772,13 @@ impl TypeChecker {
                     }
                 }
 
+                (
+                    Type::Newblob(a_name, _, a_namespace),
+                    Type::Newblob(b_name, _, b_namespace),
+                ) if a_name == b_name && a_namespace == b_namespace => {
+                    // Need not recurse - we already know the types are identical.
```
[the remark this answers is not in the document]
Comment (iwikal) [reply sylt-compiler/src/typechecker.rs:2068]: They don't look the same anymore

File: sylt-compiler/src/typechecker.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (iwikal) [reply sylt-compiler/src/typechecker.rs]: I decided to fail early instead.

File: sylt-compiler/src/typechecker.rs [hunk 5]
```diff
@@ -2032,6 +2061,13 @@ impl TypeChecker {
                     }
                 }
 
+                (
+                    Type::Newblob(a_name, _, a_args, a_namespace),
+                    Type::Newblob(b_name, _, b_args, b_namespace),
+                ) if a_name == b_name && a_namespace == b_namespace && a_args == b_args => {
+                    // Need not recurse - we already know the types are identical.
```
Comment (FredTheDino) [inline sylt-compiler/src/typechecker.rs:2068]: Args should be unified...

File: tests/newblob/constraint.sy [hunk 6]
```diff
@@ -0,0 +1,17 @@
+Foo :: newblob {
```
Comment (FredTheDino) [inline tests/newblob/constraint.sy:1]: Add a test where the newblob is passed as arguments. One as generic, and one where the type is specified.

**Review by FredTheDino — APPROVED** [review]
LGTM, but some more tests would be nice since it's a lot of code.
